neutone_sdk/core.py

```diff
@@ -37,53 +37,94 @@ class CoreMetadata(NamedTuple):
         Creates an Neutone model, wrapping a child model (that does the real
         work).
         """
         super().__init__()
-        self.MAX_N_PARAMS = constants.MAX_N_PARAMS
+        self.MAX_N_PARAMS = self._get_max_n_params()
         self.SDK_VERSION = constants.SDK_VERSION
         self.CURRENT_TIME = time.time()
         self.use_debug_mode = use_debug_mode
+        self.n_neutone_parameters = len(self.get_neutone_parameters())
 
-        assert len(self.get_neutone_parameters()) <= self.MAX_N_PARAMS
+        # Ensure the number of parameters is within the allowed limit
+        assert self.n_neutone_parameters <= self.MAX_N_PARAMS, (
+            f"Number of parameters ({self.n_neutone_parameters}) exceeds the maximum "
+            f"allowed ({self.MAX_N_PARAMS})."
+        )
         # Ensure parameter names are unique
         assert len(set([p.name for p in self.get_neutone_parameters()])) == len(
             self.get_neutone_parameters()
         )
-        model.eval()
-        self.model = model
 
-        # Convert neutone_parameters to metadata format
-        neutone_parameters = self.get_neutone_parameters()
-        if len(neutone_parameters) < self.MAX_N_PARAMS:
-            neutone_parameters += [
-                NeutoneParameter(
-                    name="",
-                    description="",
-                    used=False,
-                    default_value=0.0,
-                )
-            ] * (self.MAX_N_PARAMS - len(neutone_parameters))
+        # Save parameter metadata
         self.neutone_parameters_metadata = {
-            f"p{idx + 1}": neutone_parameter.to_metadata_dict()
-            for idx, neutone_parameter in enumerate(neutone_parameters)
+            f"p{idx + 1}": p.to_metadata_dict()
+            for idx, p in enumerate(self.get_neutone_parameters())
         }
-        default_param_values = tr.tensor(
-            [
-                neutone_parameter.default_value
-                for neutone_parameter in neutone_parameters
-            ]
-        ).unsqueeze(-1)
+
+        # Allocate default params buffer to prevent dynamic allocations later
+        numerical_default_param_vals = self._get_numerical_default_param_values()
+        default_param_values_t = tr.tensor([v for _, v in numerical_default_param_vals])
+        assert default_param_values_t.size(0) <= self.MAX_N_PARAMS, (
+            f"Number of default parameter values ({default_param_values_t.size(0)}) "
+            f"exceeds the maximum allowed ({self.MAX_N_PARAMS})."
+        )
+        default_param_values = default_param_values_t.unsqueeze(-1)
         self.register_buffer("default_param_values", default_param_values)
+
+        # Allocate remapped params dictionary to prevent dynamic allocations later
         self.remapped_params = {
-            neutone_param.name: default_param_values[idx]
-            for idx, neutone_param in enumerate(self.get_neutone_parameters())
+            name: tr.tensor([val])
```

Comment:
Just wondering why you decided to go for this approach instead of indexing in `default_param_values` like before.

Reply:
No reason just prob thought it was a little cleaner like this (I think avoiding indexing when possible is usually desirable since it can cause out of range errors)